fixed sudo checks

rcmdnk opened the pull request:
To check `sudo` command,
it is better to check `/bin/mkdir -p #{HOMEBREW_PREFIX_DEFAULT}"` command instead of using `-v`,
because users can have limited sudo right and they can `sudo -v` test.

In addition **Select the Linuxbrew installation directory** prompts were not working before.
This PR changes to do:

* Check `/usr/bin/sudo -n /bin/mkdir -p #{HOMEBREW_PREFIX_DEFAULT}`
* If not, ask user to select (Enter password for /home/linuxbrew, C-D for $HOME/, C-C to interrupt)
* If Password is entered, try `/usr/bin/sudo /bin/mkdir -p #{HOMEBREW_PREFIX_DEFAULT}`
* If not, or C-D, use $HOME/

sjackman commented on install:
```diff
@@ -253,13 +253,17 @@ unless mac?
 not intentional, please restore the default permissions and try running the
 installer again:
-    sudo chmod 775 #{HOMEBREW_PREFIX}
+    sudo chmod 775 $USER #{HOMEBREW_PREFIX}
```
What's the purpose of this change?

rcmdnk replied:
Sorry, this is just a misunderstanding that I saw somehow `chmod` as `chown`.
`$USER` was removed.

sjackman commented:
I'm happy with the purpose of this PR. Thanks for this work, @rcmdnk. The patch seems bigger than necessary though. Can we instead replace `/usr/bin/sudo -v` with `/usr/bin/sudo -l /bin/mkdir >/dev/null` ?

sjackman commented on install:
```diff
@@ -218,10 +211,6 @@ def shell_profile
-# Invalidate sudo timestamp before exiting (if it wasn't active before).
-Kernel.system "/usr/bin/sudo -n -v 2>/dev/null"
-at_exit { Kernel.system "/usr/bin/sudo", "-k" } unless $CHILD_STATUS.success?
-
```
Instead of moving this chunk let's patch line 222.

rcmdnk replied:
As `sudo` is used only when it is needed, it should be added below.

sjackman replied:
Let's solve one issue per PR. This issue need not be address in this PR.

rcmdnk replied:
This is directly related to `sudo` check below.
If a password is required, this simple check fails and does not implement `sudo -k` at exit.

sjackman commented on install:
```diff
@@ -218,10 +211,6 @@ def shell_profile
 end
 
-# Invalidate sudo timestamp before exiting (if it wasn't active before).
-Kernel.system "/usr/bin/sudo -n -v 2>/dev/null"
```
You can prevent the prompt by disabling this line on Linux.
```ruby
Kernel.system "/usr/bin/sudo -n -v 2>/dev/null" if mac?
```

rcmdnk replied:
It is unnecessary because `sudo` will be checked only when it is needed in below.

sjackman replied:
Yep, so disable this line on Linux as above.

rcmdnk replied:
On Mac, in where `sudo` requires a password,
`sudo -n -v` fails.
Therefore these lines are meaningless `if mac` is added.

sjackman replied:
These lines are from Homebrew/install on macOS.
See https://github.com/Homebrew/install/blob/8e6decdff581a4a463d66c363831f8da6b865ff8/install#L148-L150

rcmdnk replied:
Sorry, the previous comment is wrong and I had misunderstandings.
This `at_exit` code works as expected and do not needed to be moved,
as `sudo -k` do nothing as you mentioned.

I will change them back as before about `at_exit`.

sjackman commented on install:
```diff
@@ -233,18 +222,29 @@ abort <<-EOABORT unless !mac? || `dsmemberutil checkmembership -U "#{ENV["USER"]
+    out = io.read
+    io.close
+    @have_sudo = $CHILD_STATUS.success?
+    if (not @have_sudo) && out == "sudo: a password is required\n"
```
How about…
```ruby
    Kernel.system "/usr/bin/sudo -n -l /bin/mkdir >/dev/null 2>/dev/null"
    unless $CHILD_STATUS.success?
…
```

rcmdnk replied:
Some users have right to do `sudo` with password, and some have right w/o password.
This is needed to distinguish users which do not have the right anyway.
Otherwise, the next prompt will be strange for these users.

rcmdnk replied:
`sudo -l /bin/mkdir` instead of `sudo /bin/mkdir -p #{HOMEBREW_PREFIX_DEFAULT}` seems fine.

sjackman replied:
You can use `sudo -nv` to check whether a user has `sudo` access without a password.

rcmdnk replied:
Yes, but it fails if the user can do `sudo` only with a password.
Therefore, here, 

1. first check if the user can use `sudo` without password by `sudo -n -l mkdir`.
2. If failed, check if a password is required or not by the output.
3. Only if a password is required, ask user to put password to create /home/linuxbrew otherwise make $HOME/.linuxbrew

sjackman replied:
I suggest…
```ruby
if `sudo -nv 2>&1` == "sudo: a password is required\n"
```

rcmdnk replied:
Yes, that is enough in this case. updated.

sjackman commented on install:
```diff
@@ -253,13 +253,17 @@ unless mac?
+# Invalidate sudo timestamp before exiting (if it wasn't active before).
+at_exit { Kernel.system "/usr/bin/sudo", "-k" } if @have_sudo
+
+
```
Please remove this chunk.

rcmdnk replied:
This needs to be here because `have_sudo` is just set above.

sjackman replied:
I believe `sudo -k` is successful whether or not the user has `sudo` access, so it doesn't need to be conditional on `@have_sudo`.

rcmdnk replied:
Yes, both cases work in the same as a result.

I just realized that if we want to implement correctly as a comment **(if it wasn't active before)**,
we put this line after line 242.

sjackman commented on install-ruby:
```diff
@@ -9,9 +9,8 @@ case $(uname -m) in
 esac
 
-if test -w /home/linuxbrew || sudo -v; then
+if test -w /home/linuxbrew || sudo mkdir -p /home/linuxbrew; then
```
I like this change. Thanks! Could you please add `sudo -p` to add a custom prompt?
```suggestion
if test -w /home/linuxbrew || /usr/bin/sudo -p "[sudo] Enter password for $USER to install Ruby: " mkdir -p /home/linuxbrew; then
```

rcmdnk replied:
This seems good, included.

sjackman reviewed (changes requested):
Thanks for this work!

sjackman commented:
Please fix these style errors:
```
$ gem install rubocop
$ rubocop
Inspecting 3 files
C..
Offenses:
install:236:9: C: Style/Not: Use ! instead of not.
    if (not @have_sudo) && out == "sudo: a password is required\n"
        ^^^
install:244:7: C: Metrics/BlockNesting: Avoid more than 3 levels of block nesting.
      rescue Interrupt ...
      ^^^^^^^^^^^^^^^^
3 files inspected, 2 offenses detected
The command "rubocop" exited with 1.
```
See the TravisCI output https://travis-ci.org/Linuxbrew/install/builds/490450955

rcmdnk commented:
To reduce block nests, a new function `prefix` was implemented.

sjackman commented on install:
```diff
@@ -91,6 +90,37 @@ def system(*args)
+    return "#{ENV["HOME"]}/.linuxbrew"
+  end
+end
+
```
Please avoid large diff changes.

sjackman replied:
They are difficult to review.

rcmdnk committed:
Fixed sudo checks

Use mkdir to check sudo in install-ruby.
Use mkdir to check sudo in install, for other than initial validation check.
Ask password only when sudo failed by password requirement.
Trap Ctrl-C at sudo.

rcmdnk commented:
I refactored the code and commits.
I think this is a minimum change to have similar functions as before.

sjackman commented on install:
```diff
@@ -94,8 +94,12 @@ end
-  @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success?
+  begin
+    Kernel.system "/usr/bin/sudo", "-l", "mkdir"
+    @have_sudo = $CHILD_STATUS.success?
```
```suggestion
    @have_sudo = $CHILD_STATUS&.success?
```
`$CHILD_STATUS` can be `nil` if `/usr/bin/sudo` does not exist.

rcmdnk replied:
Changed to 

    @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success? || false

to keep compatibility for Ruby 2.2

sjackman replied:
No need for the `|| false`. The falsey value of `nil` for false is fine.

sjackman commented on install:
```diff
@@ -94,8 +94,12 @@ end
+    @have_sudo = $CHILD_STATUS.success?
+  rescue Interrupt
+    exit!
+  end
```
Is the `begin/rescue/end` block needed? Please remove it if not.

rcmdnk replied:
If this does not exist,
Ctrl-C shows traceback message.

sjackman replied:
Thanks for the explanation. I believe `exit` should work and also prevent the traceback. Let's use that rather than `exit!`.

rcmdnk replied:
I think I misunderstood your comment.
`begin`~`end` is not necessary here, so I removed it.

sjackman commented on install:
```diff
@@ -237,8 +241,8 @@ unless mac?
-    Kernel.system "/usr/bin/sudo -n -v 2>/dev/null"
-    unless $CHILD_STATUS.success?
+    out = `/usr/bin/sudo -n -l mkdir 2>&1`
+    if !$CHILD_STATUS.success? && out == "sudo: a password is required\n"
```
Could you please rename `out` to `sudo_output`?

rcmdnk replied:
Renamed

sjackman commented:
This PR is getting much closer. Thank you, @rcmdnk!

sjackman commented:
If you know how to squash a PR down to a single commit using `git rebase -i origin/master`, please do that. If you have no idea what I'm talking about, don't worry about it. I'll squash it when I merge it.

rcmdnk commented:
Yes, these commits were squished.

sjackman commented:
@rcmdnk Are you interested in continuing with this PR?

rcmdnk commented:
Yes, it should be merged otherwise the script does not work in some of my nodes.

I squished commits in a single commit.
Do you need any other changes to merge?

sjackman commented:
Yes, there's a number of unresolved comments.
See https://github.com/Linuxbrew/install/pull/47/files

rcmdnk commented:
Sorry, I missed these comments.
I updated a commit for them.

sjackman commented on install:
```diff
@@ -94,8 +94,12 @@ end
+    Kernel.system "/usr/bin/sudo", "-l", "mkdir"
+    @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success? || false
+  rescue Interrupt
+    exit!
```
```suggestion
    exit
```
`exit` also prevents the traceback.

rcmdnk replied:
As `SystemExit` trap does not exist, they work almost the same.
But maybe `exit` is better for the future update.

The difference here is just a default exit code (1 or 0).
I thought it should be an error exit, but you can decide later if necessary.

I changed it to `exit`.

sjackman replied:
Good point. `exit 1` is probably better.
```suggestion
    exit 1
```

sjackman commented on install:
```diff
@@ -94,8 +94,12 @@ end
-  @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success?
+  begin
+    Kernel.system "/usr/bin/sudo", "-l", "mkdir"
+    @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success? || false
```
```suggestion
    @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success?
```
`nil` for `false` is fine.

sjackman replied:
```suggestion
    @have_sudo = $CHILD_STATUS.success?
```
I was wrong here. `$CHILD_STATUS` is not `nil` when `/usr/bin/sudo` does not exist.

rcmdnk replied:
Maybe I also did a similar misunderstanding that I observed `nil` before adding `require "English" in the test script...
The last one is enough, I agree.

I put `|| false` because if `@have_sudo` is nil, `sudo?` cannot pass the first line always.

Anyways, I updated to just `$CHILD_STATUS.success?`.

sjackman replied:
> Maybe I also did a similar misunderstanding that I observed nil before adding `require "English" in the test script...

I think I made the identical error.

> I put || false because if @have_sudo is nil, sudo? cannot pass the first line always.

Yes, good point. I missed that.

sjackman commented on install:
```diff
@@ -94,8 +94,10 @@ end
+  Kernel.system "/usr/bin/sudo", "-l", "mkdir"
+  @have_sudo = $CHILD_STATUS.success?
+rescue Interrupt
+  exit
```
```suggestion
  exit 1
```

sjackman commented:
Thank you very much for these improvements, @rcmdnk ! And thank you for your first contribution to Linuxbrew/install! 🏅
